subsys/bluetooth/host/gatt.c

```diff
@@ -3822,6 +4003,70 @@ static int ccc_set(const char *name, size_t len_rd, settings_read_cb read_cb,
 
 SETTINGS_STATIC_HANDLER_DEFINE(bt_ccc, "bt/ccc", NULL, ccc_set, NULL, NULL);
 
+#if defined(CONFIG_BT_GATT_SERVICE_CHANGED)
+static int sc_set(const char *name, size_t len_rd, settings_read_cb read_cb,
+		  void *cb_arg)
+{
+	struct gatt_sc_cfg *cfg;
+	u8_t id;
+	bt_addr_le_t addr;
+	int len, err;
+	const char *next;
+
+	if (!name) {
+		BT_ERR("Insufficient number of arguments");
+		return -EINVAL;
+	}
+
+	err = bt_settings_decode_key(name, &addr);
+	if (err) {
+		BT_ERR("Unable to decode address %s", log_strdup(name));
+		return -EINVAL;
+	}
+
+	settings_name_next(name, &next);
+
+	if (!next) {
+		id = BT_ID_DEFAULT;
+	} else {
+		id = strtol(next, NULL, 10);
+	}
+
+	cfg = find_sc_cfg(id, &addr);
+	if (!cfg && len_rd) {
+		/* Find and initialize a free sc_cfg entry */
+		cfg = find_sc_cfg(BT_ID_DEFAULT, BT_ADDR_LE_ANY);
+		if (!cfg) {
+			BT_ERR("Unable to restore SC: no cfg left");
+			return 0;
+		}
+
+		cfg->id = id;
+		bt_addr_le_copy(&cfg->peer, &addr);
+	}
+
+	if (len_rd) {
+		len = read_cb(cb_arg, cfg->data, sizeof(cfg->data));
+		if (len < 0) {
+			BT_ERR("Failed to decode value (err %d)", len);
+			return len;
+		}
+		BT_DBG("Read SC: len %d", len);
+
+		BT_DBG("Restored SC for %s", bt_addr_le_str(&addr));
+	} else if (cfg) {
+		/* Clear configuration */
+		memset(cfg->data, 0, sizeof(cfg->data));
+
+		BT_DBG("Removed SC for %s", bt_addr_le_str(&addr));
```

Comment:
Shouldn't you also be doing this:
```
cfg->id = BT_ID_DEFAULT;
 bt_addr_le_copy(&cfg->peer, BT_ADDR_LE_ANY);
```

At least you seem to be doing that in the `sc_clear()` function, and this situation is essentially the same.

Reply:
Fixed in latest push, I changed it to use sc_clear, as we also need to persist the data.

Reply:
Hmm? I don't see that change. Also, if you did do it I think it wouldn't be correct, since here you're reading data from flash, and trying to write something back to flash by calling `sc_clear()` would create a sort of loopback.

Reply:
Sorry I misread the line number in your comment. I thought you were talking about the memset(cfg->data, 0, sizeof(cfg->data)); in the sc_restore function, where I know call sc_clear.

Just pushed a fix to this one as well in sc_set(...).